src/sqlfluff/rules/L003.py

```diff
@@ -567,207 +661,171 @@ def _process_current_line(
-                                                tab_space_size=self.tab_space_size,
-                                            ),
-                                        ),
-                                    ],
-                                ),
-                            ],
-                        )
-                elif this_indent_num < comp_indent_num:
+                return LintResult(
+                    anchor=trigger_segment,
+                    memory=memory,
+                    description=(
+                        "Indentation not hanging or a multiple of {} spaces"
+                    ).format(self.tab_space_size),
+                    fixes=fixes,
+                )
+
+            # We'll need this value later.
+            this_indent_num = this_line["indent_size"] // self.tab_space_size
+
+            # We know that the indent balance is higher, what actually is
+            # the difference in indent counts? It should be a whole number
+            # if we're still here.
+            comp_indent_num = prev_line["indent_size"] // self.tab_space_size
+
+            indents = this_line_segments.select(sp.is_type("indent"))
+            positive_indents = indents.select(select_if=sp.not_(sp.is_type("dedent")))
+            start_line_indents = this_line_segments.select(
+                select_if=sp.and_(sp.is_type("indent"), sp.not_(sp.is_type("dedent"))),
+                stop_seg=this_line_segments.first(sp.is_code()).get(0),
+            )
+            dedents = indents.select(
+                select_if=sp.is_type(
+                    "dedent",
+                    # I believe the below 2 are not required in the new algo
+                    # "end_bracket",
+                    # "end_square_bracket",
+                )
+            )
+            # The indent number should be at least 1, and can be UP TO
+            # and including the difference in the indent balance.
+            if comp_indent_num == this_indent_num:
+                # We have two lines indented the same, but with a different starting
+                # indent balance. This is either a problem OR a sign that one of the
+                # opening indents wasn't used. We account for the latter and then
+                # have a violation if that wasn't the case.
+
+                # Does the comparison line have enough unused indent to get us back
+                # to where we need to be? NB: This should only be applied if this is
+                # a CLOSING bracket.
+
+                # First work out if we have some closing brackets, and if so, how
+                # many.
+                # Due to strange inheritence all dedents are also indents
+
+                self.logger.debug((positive_indents, dedents, start_line_indents))
+                # Calc the balance of this line.
+                # Double the contribution for any indents at the start of the line.
```

Comment:
Poorly expressed and WIP. Also I'm trying to reuse the existing logic to achieve something that it might not be able. 
Basically we are trying to see if during this line we balance out for example do we have a double indent somewhere and double decent here making this actually balanced. However sometimes we have lines essentially starting with indents which are requesting that we push this line deeper. I think an actually stack of pairing would solve this

Reply:
+1 for leveraging good data structures where applicable. I think this code was prototyped into existence, and sometimes the patterns are only obvious in retrospect.